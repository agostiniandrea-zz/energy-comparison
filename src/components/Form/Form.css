.form-container {
    border: 3px solid green;
    border-radius: 15px;
}

.form-container .invalid-value {
    visibility: hidden;
}

.form-container.show-invalid-parameters .invalid-value {
    visibility: visible;
}

.form-item {
    min-height: 50px;
    text-align: left;
    margin-bottom: 15px;
}

.form-item.email input {
    width: 95%;
}

.form-label,
.invalid-value {
    font-weight: bold;
}

.form-label {
    font-size: 16px;
    display: block;
    margin-bottom: 5px;
}

.invalid-value {
    font-size: 12px;
    position: absolute;
    margin-top: 3px;
    margin-left: 5px;
    color: red;
}

.radio-button {
    font-size: 14px;
    margin-bottom: 10px;
}

.radio-button input {
    position: absolute;
    margin-left: 5px;
}

.usage-unity {
    margin-left: 5px;
    color: grey;
    font-size: 12px;
}

/* Smartphones (portrait and landscape) ----------- */
@media only screen and (min-device-width : 320px) {
    .form-container {
        display: grid;
        padding: 10px;
        margin-top: 20px;
        width: auto;
    }

    .submit-button {
        margin-top: 10px;
        width: 150px;
        height: 50px;
        border-color: red;
        background: red;
        color: white;
        font-weight: bold;
        font-size: 12px;
    }
}

/* iPads (portrait and landscape) ----------- */
@media only screen and (min-device-width : 768px) {
    .form-container {
        display: flex;
        margin: 40px auto 0;
    }

    .form-item {
        min-width: 125px;
        margin-right: 10px;
    }

    .form-item.energy-type {
        min-width: 110px;
    }

    .submit-button {
        width: 100px;
        min-height: 30px;
        margin-left: 10px;
        margin-top: 27px;
    }
}

@media only screen and (min-width : 1024px) {
    .form-container {
        margin: 60px auto 0;
        padding: 30px;
    }

    .form-item {
        width: 150px;
    }

    .form-item.email {
        width: 180px;
        margin-right: 30px;
    }

    .submit-button {
        margin-left: 40px;
    }
}

@media only screen and (min-width : 1366px) {
    .form-item {
        width: 200px;
    }

    .submit-button {
        margin-left: 100px;
    }
}