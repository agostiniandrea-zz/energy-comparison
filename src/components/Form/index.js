import React from 'react';
import './Form.css';

export const regex = {
    postcode: /^[1-9][0-9]{3} ?(?!sa|sd|ss)[a-z]{2}$/i,
    email: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
};

const typeValues = {
    "Electricity": 0,
    "Gas": 1,
}

const energyUnities = ["kWh", "m3"];

class Form extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            postcode: {
                value: "",
                isValid: false,
            },
            email: {
                value: "",
                isValid: false,
            },
            type: {
                value: 0,
                isValid: true,
            },
            usage: {
                value: 0,
                isValid: false,
            },
            showInvalidParameters: false
        };

        this.calculate = this.calculate.bind(this);
        this.onPostcodeChange = this.onPostcodeChange.bind(this);
        this.onEmailChange = this.onEmailChange.bind(this);
        this.onTypeChange = this.onTypeChange.bind(this);
        this.onUsageChange = this.onUsageChange.bind(this);
    }

    onPostcodeChange(event) {
        const postcode = event.target.value.toUpperCase();
        this.saveStateChange("postcode", postcode, regex.postcode.test(postcode));
    }

    onEmailChange(event) {
        this.saveStateChange("email", event.target.value, regex.email.test(this.state.email.value));
    }

    onTypeChange(value) {
        this.saveStateChange("type", value);
        this.saveUsageChange(0);
        this.props.onTypeChange();
    }

    saveStateChange(type, fieldValue, isValid = true) {
        this.setState({
            [type]: {
                value: fieldValue,
                isValid
            }
        });
    }

    onUsageChange(event) {
        this.saveUsageChange(event.target.value);
    }

    saveUsageChange(value) {
        this.saveStateChange("usage", value, value > 0);
    }

    calculate(e) {
        if (!this.state.showInvalidParameters) {
            this.setState({
                showInvalidParameters: true
            });
        }
        if (this.state.postcode.isValid && this.state.email.isValid && this.state.usage.isValid && this.state.type.isValid) {
            this.props.onSubmit(this.state);
        }
        if (e) {
            e.preventDefault();
        }
    }

    render() {
        return (
            <form className={`form-container ${this.state.showInvalidParameters ? "show-invalid-parameters" : ""}`}>
                <div className="form-item postcode">
                    <label className="form-label">Postal code</label>
                    <input
                        type="text"
                        value={this.state.postcode.value}
                        onChange={this.onPostcodeChange}
                        maxLength={6}
                        placeholder="1234AB"
                    />
                    {!this.state.postcode.isValid && <label className="invalid-value">X</label>}
                </div>
                <div className="form-item energy-type">
                    <label className="form-label">Energy type</label>
                    <div className="radio-button">
                        <label>Electricity</label>
                        <input
                            type="radio"
                            checked={this.state.type.value === typeValues["Electricity"]}
                            onChange={() => this.onTypeChange(typeValues["Electricity"])}
                            value={typeValues["Electricity"]}
                        />
                    </div>
                    <div className="radio-button">
                        <label>Gas</label>
                        <input
                            type="radio"
                            checked={this.state.type.value === typeValues["Gas"]}
                            onChange={() => this.onTypeChange(typeValues["Gas"])}
                            value={typeValues["Gas"]}
                        />
                    </div>
                </div>
                <div className="form-item email">
                    <label className="form-label">Email address</label>
                    <input
                        type="email"
                        value={this.state.email.value}
                        onChange={this.onEmailChange}
                        placeholder="your@email.com"
                    />
                    {!this.state.email.isValid && <label className="invalid-value">X</label>}
                </div>
                <div className="form-item energy-usage">
                    <label className="form-label">Usage</label>
                    <input
                        type="number"
                        value={this.state.usage.value}
                        onChange={this.onUsageChange}
                        placeholder={this.state.type.value}
                    />
                    <label className="usage-unity">{energyUnities[this.state.type.value]}</label>
                    {!this.state.usage.isValid && <label className="invalid-value">X</label>}
                </div>
                <button className="submit-button" onClick={this.calculate}>Start comparing</button>
            </form>
        );
    }
}

export default Form;
