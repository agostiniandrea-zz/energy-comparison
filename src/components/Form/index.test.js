import React from 'react';
import { shallow } from 'enzyme';
import Form from './index';
import { regex } from './index';

const VALID_EMAIL = "name@gmail.com";
const VALID_POSTCODE = "1059CL";
const VALID_USAGE = 100;

const INVALID_EMAIL = "namegmail.com";
const INVALID_POSTCODE = "100010";
const INVALID_USAGE = 0;

let wrapper;

beforeEach(() => {
    wrapper = shallow(<Form />);
})

afterEach(() => {
    wrapper.unmount();
});

describe('Form/index.js', () => {
    it('renders without crashing', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('shows invalid labels', () => {
        wrapper.instance().calculate();
        expect(wrapper.instance().state.showInvalidParameters).toBe(true);
        expect(wrapper.find(".postcode").find(".invalid-value").exists()).toBe(true);
        expect(wrapper.find(".email").find(".invalid-value").exists()).toBe(true);
        expect(wrapper.find(".energy-usage").find(".invalid-value").exists()).toBe(true);
    });

    it('sets valid values', () => {
        wrapper.instance().saveStateChange("postcode", VALID_POSTCODE, regex.postcode.test(VALID_POSTCODE));
        expect(wrapper.instance().state.postcode).toStrictEqual({ isValid: true, value: VALID_POSTCODE });

        wrapper.instance().saveStateChange("email", VALID_EMAIL, regex.email.test(VALID_EMAIL));
        expect(wrapper.instance().state.email).toStrictEqual({ isValid: true, value: VALID_EMAIL });

        wrapper.instance().saveUsageChange(VALID_USAGE);
        expect(wrapper.instance().state.usage).toStrictEqual({ isValid: true, value: VALID_USAGE });

        wrapper.instance().saveStateChange("type", 0);
        expect(wrapper.instance().state.type).toStrictEqual({ isValid: true, value: 0 });
    });

    it('warns on invalid values', () => {
        expect(wrapper.instance().state.usage).toStrictEqual({ isValid: false, value: 0 });

        wrapper.instance().saveStateChange("postcode", INVALID_POSTCODE, regex.postcode.test(INVALID_POSTCODE));
        expect(wrapper.instance().state.postcode).toStrictEqual({ isValid: false, value: INVALID_POSTCODE });

        wrapper.instance().saveStateChange("email", INVALID_EMAIL, regex.email.test(INVALID_EMAIL));
        expect(wrapper.instance().state.email).toStrictEqual({ isValid: false, value: INVALID_EMAIL });

        wrapper.instance().saveUsageChange(INVALID_USAGE);
        expect(wrapper.instance().state.usage).toStrictEqual({ isValid: false, value: INVALID_USAGE });
    });
});