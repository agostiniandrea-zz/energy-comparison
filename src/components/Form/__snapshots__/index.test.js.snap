// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form/index.js renders without crashing 1`] = `ShallowWrapper {}`;
